### PR TITLE
[IOPLT-578] - Process active subscription request

### PR DESCRIPTION
**Pull request by datalek**
#### List of Changes

- Rename `insertSubscription` use case to `createSubscription`
- Remove `SubscriptionRequest` type
- Rename `subscriptionRequestWriter` to `subscriptionQueue`
- Move the creation of active subscription from `createSubscription` to `processSubscriptionRequest`, from sync to async.

#### Motivation and Context

The system must update the information about activations if the system manager activates a user.

#### Checklist before requesting a review

- [ ] I have performed a self-review of my code.
- [ ] I have committed only changes relevant to the task.
- [ ] I have minimized the number of changes.
- [ ] My changes are about only one task or issue.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: apps/functions-subscription/src/adapters/azure/eventhubs/__tests__/subscription.test.ts [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+
+    mockEventHub.sendBatch.mockRejectedValueOnce(error);
+
+    const actual = await makeSubscriptionQueueEventHubProducer(
+      mockEventHub as unknown as EventHubProducerClient,
+    ).enqueue(aSubscription)();
+
+    expect(actual).toStrictEqual(E.left(error));
+  });
+});
```
Comment (datalek) [inline apps/functions-subscription/src/adapters/azure/eventhubs/__tests__/subscription.test.ts:37]: **note**: This was the `subscription-request.test.ts` file, just renamed it

File: apps/functions-subscription/src/adapters/azure/eventhubs/subscription.ts [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+): SubscriptionQueue => ({
+  enqueue: (request) => {
+    const encoded = SubscriptionCodec.encode(request);
+    return pipe(
+      TE.tryCatch(() => client.sendBatch([{ body: encoded }]), E.toError),
+      // if succeeded return the request
+      TE.map(() => request),
+    );
+  },
+});
```
Comment (datalek) [inline apps/functions-subscription/src/adapters/azure/eventhubs/subscription.ts:21]: **note**: Also this one, it was the `subscription-request.ts` file

File: apps/functions-subscription/src/use-cases/process-subscription-request.ts [hunk 3]
```diff
@@ -1,54 +1,66 @@
...
-          ),
-          TE.map(() => ({ trialId, userId })),
-          TE.orElse(recoverItemAlreadyExists(subscription)),
-        ),
+    RTE.bindW('activationRequest', () =>
+      pipe(
+        makeInsertActivationRequest({
+          trialId: subscription.trialId,
+          userId: subscription.userId,
+          activated: subscription.state !== 'SUBSCRIBED',
```
Comment (kin0992) [inline apps/functions-subscription/src/use-cases/process-subscription-request.ts:55]: **question**: shouldn't be this `subscription.state === 'ACTIVE'`?

File: apps/functions-subscription/src/use-cases/process-subscription-request.ts [hunk 4]
```diff
@@ -1,54 +1,66 @@
...
-            env.subscriptionHistoryWriter.insert(subscriptionHistory),
-          ),
-          TE.orElse(recoverItemAlreadyExists(subscription)),
-          TE.flatMap(() =>
-            env.activationRequestWriter.insert(activationRequest),
-          ),
-          TE.map(() => ({ trialId, userId })),
-          TE.orElse(recoverItemAlreadyExists(subscription)),
-        ),
+    RTE.bindW('activationRequest', () =>
```
Comment (kin0992) [inline apps/functions-subscription/src/use-cases/process-subscription-request.ts:50]: **note**: I think you can avoid the use of `bind`.

In this case, the binded values are not used anywhere else. I guess you can remove them and use few `RTE.flatMap` when there are the `RTE.bindW` methods.

File: apps/functions-subscription/src/use-cases/create-subscription.ts [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+          O.fold(
+            () => handleMissingSubscription(subscription),
+            () =>
+              RTE.left(new ItemAlreadyExists('Subscription already exists')),
+          ),
+        ),
+      ),
+    ),
+    RTE.map(({ subscription }) => subscription),
+  );
```
Comment (kin0992) [inline apps/functions-subscription/src/use-cases/create-subscription.ts:50]: **note**: you can remove the `RTE.Do` and `RTE.apS`.

```suggestion
  pipe(
    makeSubscription({ userId, trialId, state }),
    RTE.flatMap((subscription) =>
      pipe(
        getSubscriptionById(subscription.id),
        RTE.flatMap(
          O.fold(
            () => handleMissingSubscription(subscription),
            () =>
              RTE.left(new ItemAlreadyExists('Subscription already exists')),
          ),
        ),
      ),
    ),
    RTE.map((subscription) => subscription),
  );
```

File: apps/functions-subscription/src/use-cases/process-subscription-request.ts (hunk 3 above)
Comment (datalek) [reply apps/functions-subscription/src/use-cases/process-subscription-request.ts]: Yes, it can. I've added some tests and logic to handle it properly

**Review by kin0992 — APPROVED** [review]

**Pull request merged**